aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -0,0 +1,896 @@
+    bool fn_train, bool fn_bidirectional, IntList fn_batch_sizes,
+    const Tensor& fn_dropout_state
+    ) {
+
+  auto input = input_r;
+  auto weight_buf = weight_buf_r;
+
+  RNNParams fn;
+  fn.rnn.set_mode(fn_mode);
+  fn.rnn.hidden_size = fn_hidden_size;
+  fn.rnn.num_layers = fn_num_layers;
+  fn.rnn.set_bidirectional(fn_bidirectional);
+  fn.rnn.datatype = getCudnnDataType(input);
+  fn.dropout.set(fn_train, fn_dropout, fn_dropout_state);
+  fn.tensors.set(input.sizes(), fn_batch_sizes, batch_first);
+
+  // TODO: Set device to input
+
+  if (fn.rnn.mode != CUDNN_LSTM) {
+    if (cx.defined()) {
+      throw std::runtime_error("rnn: illegal defined cx for non-LSTM RNN");
+    }
+  }
+
+  // TODO: can batch_first be a wrapper around this function?
+  auto is_input_packed = fn.tensors.batch_sizes.size() != 0;
+  if (batch_first && !is_input_packed) {
+    input = input.transpose(0, 1);
+  }
+
+  auto hidden_size = _hidden_size(fn.rnn, fn.tensors);
+  auto output_size = _output_size(fn.rnn, fn.tensors);
+
+  AT_ASSERT(hx.is_contiguous(), "hx.is_contiguous()");
+  AT_ASSERT(!cx.defined() || cx.is_contiguous(), "!cx or cx.is_contiguous()");
+
+  auto x = input.contiguous();
+  auto output = input.type().tensor(output_size);
+  auto hy = hx.type().tensor(hidden_size);
+  Tensor cy;
+  if (cx.defined()) {
+    cy = cx.type().tensor(hidden_size);
+  } else {
+    cy = hx.type().tensor(); // NB: Not allowed to return undefined tensors
+  }
+  auto y = output;
+
+  auto handle = getCudnnHandle();
+  RNNDescriptors descs(fn, handle, x, y, hx, cx);
+
+  FilterDescriptor w_desc;
+  if (!weight_buf.defined()) {
+    auto num_weights = get_num_weights(handle, descs.rnn_desc, descs.x_descs[0], fn.rnn.datatype);
+    weight_buf = x.type().tensor(num_weights);
+    w_desc.set(weight_buf, 3);
+    weight_buf.zero_();
+    std::vector<Tensor> params;
+    size_t params_stride0;
+    std::tie(params, params_stride0) = get_parameters(handle, fn.rnn, descs.rnn_desc, descs.x_descs[0], w_desc, weight_buf);
+    _copyParams(MatrixRef<Tensor>{weight, static_cast<size_t>(weight_stride0)},
+                MatrixRef<Tensor>{params, params_stride0});
+  } else {
+    w_desc.set(weight_buf, 3);
+  }
+
+  if (cx.defined() && !cx.sizes().equals(hidden_size)) {
+    std::ostringstream oss;
+    oss << "Expected cell size " << IntList{hidden_size} << ", got " << cx.sizes();
+    throw std::runtime_error(oss.str());
+  }
+
+  size_t workspace_size;
+  auto x_descs_arr = descs.get_x_descs();
+  auto y_descs_arr = descs.get_y_descs();
+  CUDNN_CHECK(cudnnGetRNNWorkspaceSize(
+        handle,
+        descs.rnn_desc.desc(),
+        fn.tensors.seq_length,
+        x_descs_arr.data(),
+        &workspace_size
+        ));
+  Tensor workspace = input.type().toScalarType(kByte).tensor(workspace_size);
+
+  Tensor reserve;
+  // NB: Previously, the test was for fn.requires_grad, but we don't have
+  // this information.  Use 'train' as a proxy.
+  if (fn.dropout.train) {
```

Comment:
Why do we have to look into dropout properties to figure out if we're in training mode or not? I would move this property to the top level of `fn` (it doesn't only affect the dropout, but potentially the algo as can be seen in this part of code).

Reply:
I'll `s/fn.dropout.train/fn_train/` then.